File: sdk/go/queue_control.go
package bunqueue

import "strings"

// Pause stops delivery for this queue (producers can still add).
func (q *Queue) Pause() error {
	_, err := q.call(map[string]any{"cmd": "Pause", "queue": q.Name})
	return err
}

// Resume re-enables delivery for this queue.
func (q *Queue) Resume() error {
	_, err := q.call(map[string]any{"cmd": "Resume", "queue": q.Name})
	return err
}

// IsPaused reports the queue's paused flag (top-level `paused`).
func (q *Queue) IsPaused() (bool, error) {
	response, err := q.call(map[string]any{"cmd": "IsPaused", "queue": q.Name})
	if err != nil {
		return false, err
	}
	return asBool(response["paused"]), nil
}

// Drain removes all waiting/delayed jobs and returns how many were dropped.
func (q *Queue) Drain() (int, error) {
	response, err := q.call(map[string]any{"cmd": "Drain", "queue": q.Name})
	if err != nil {
		return 0, err
	}
	return asInt(response["count"]), nil
}

// Clean removes jobs in `state` older than graceMs and returns their ids.
func (q *Queue) Clean(graceMs, limit int, state string) ([]string, error) {
	response, err := q.call(map[string]any{
		"cmd": "Clean", "queue": q.Name, "grace": graceMs, "limit": limit, "state": state,
	})
	if err != nil {
		return nil, err
	}
	return asStrings(response["ids"]), nil
}

// Obliterate wipes the queue entirely (jobs, DLQ, metadata).
func (q *Queue) Obliterate() error {
	_, err := q.call(map[string]any{"cmd": "Obliterate", "queue": q.Name})
	return err
}

// Remove cancels a job by id; false when it no longer exists.
func (q *Queue) Remove(id string) (bool, error) {
	_, err := q.call(map[string]any{"cmd": "Cancel", "id": id})
	if err != nil {
		if isNotFound(err) {
			return false, nil
		}
		return false, err
	}
	return true, nil
}

// Discard marks a job so it will not be retried again.
func (q *Queue) Discard(id string) error {
	_, err := q.call(map[string]any{"cmd": "Discard", "id": id})
	return err
}

// Promote moves a delayed job to waiting immediately.
func (q *Queue) Promote(id string) error {
	_, err := q.call(map[string]any{"cmd": "Promote", "id": id})
	return err
}

// RetryJob re-queues a failed/completed job (server command: MoveToWait).
func (q *Queue) RetryJob(id string) error {
	_, err := q.call(map[string]any{"cmd": "MoveToWait", "id": id})
	return err
}

// ChangePriority updates a waiting job's priority.
func (q *Queue) ChangePriority(id string, priority int) error {
	_, err := q.call(map[string]any{"cmd": "ChangePriority", "id": id, "priority": priority})
	return err
}

// ChangeDelay updates a delayed job's remaining delay.
func (q *Queue) ChangeDelay(id string, delayMs int) error {
	_, err := q.call(map[string]any{"cmd": "ChangeDelay", "id": id, "delay": delayMs})
	return err
}

// UpdateJobData replaces a job's user data.
func (q *Queue) UpdateJobData(id string, data map[string]any) error {
	_, err := q.call(map[string]any{"cmd": "Update", "id": id, "data": data})
	return err
}

// MoveJobToFailed fails an ACTIVE job explicitly, mirroring the worker's FAIL
// wire: the stack and the unrecoverable flag travel too, not just the message
// (#111 silent-loss class). The job must have been pulled: pass its token.
func (q *Queue) MoveJobToFailed(id string, failure error, token string) error {
	message := failure.Error()
	// A Go error carries no stack; keep the message line (the server persists
	// the FIRST stackTraceLimit lines, and Go stacks lead with the site).
	stack := []string{errorTypeName(failure) + ": " + message}
	var unrecoverable any
	var unrec *UnrecoverableError
	if asUnrecoverable(failure, &unrec) {
		unrecoverable = true
	}
	_, err := q.call(map[string]any{
		"cmd":           "FAIL",
		"id":            id,
		"error":         message,
		"stack":         stack,
		"unrecoverable": unrecoverable,
		"token":         nonEmpty(token),
	})
	return err
}

func nonEmpty(s string) any {
	if s == "" {
		return nil
	}
	return s
}

func errorTypeName(err error) string {
	name := strings.TrimPrefix(strings.TrimPrefix(typeName(err), "*"), "bunqueue.")
	return name
}

File: sdk/rust/src/queue_control.rs
use rmpv::Value;

use crate::wire::{command, get, map};
use crate::{Queue, Result};

impl Queue {
    pub fn pause(&self) -> Result<()> {
        self.unit("Pause", map([("queue", Value::from(self.name.clone()))]))
    }

    pub fn resume(&self) -> Result<()> {
        self.unit("Resume", map([("queue", Value::from(self.name.clone()))]))
    }

    pub fn is_paused(&self) -> Result<bool> {
        let response = self.call(command(
            "IsPaused",
            map([("queue", Value::from(self.name.clone()))]),
        ))?;
        Ok(get(&response, "paused").and_then(Value::as_bool) == Some(true))
    }

    pub fn drain(&self) -> Result<i64> {
        let response = self.call(command(
            "Drain",
            map([("queue", Value::from(self.name.clone()))]),
        ))?;
        Ok(get(&response, "count").and_then(Value::as_i64).unwrap_or(0))
    }

    pub fn clean(&self, grace_ms: i64, limit: i64, state: &str) -> Result<Vec<String>> {
        let response = self.call(command(
            "Clean",
            map([
                ("queue", Value::from(self.name.clone())),
                ("grace", Value::from(grace_ms)),
                ("limit", Value::from(limit)),
                ("state", Value::from(state)),
            ]),
        ))?;
        Ok(get(&response, "ids")
            .map(crate::wire::strings)
            .unwrap_or_default())
    }

    pub fn obliterate(&self) -> Result<()> {
        self.unit(
            "Obliterate",
            map([("queue", Value::from(self.name.clone()))]),
        )
    }

    pub fn remove(&self, id: &str) -> Result<()> {
        self.unit("Cancel", map([("id", Value::from(id))]))
    }

    pub fn promote(&self, id: &str) -> Result<()> {
        self.unit("Promote", map([("id", Value::from(id))]))
    }

    pub fn retry_job(&self, id: &str) -> Result<()> {
        self.unit("MoveToWait", map([("id", Value::from(id))]))
    }

    pub fn change_priority(&self, id: &str, priority: i64) -> Result<()> {
        self.unit(
            "ChangePriority",
            map([("id", Value::from(id)), ("priority", Value::from(priority))]),
        )
    }

    pub fn change_delay(&self, id: &str, delay_ms: i64) -> Result<()> {
        self.unit(
            "ChangeDelay",
            map([("id", Value::from(id)), ("delay", Value::from(delay_ms))]),
        )
    }

    pub fn update_job_data(&self, id: &str, data: Value) -> Result<()> {
        self.unit("Update", map([("id", Value::from(id)), ("data", data)]))
    }

    pub(crate) fn unit(&self, name: &str, fields: crate::wire::Map) -> Result<()> {
        self.call(command(name, fields))?;
        Ok(())
    }
}
